main.go

```diff
@@ -88,6 +108,48 @@ func main() {
 		os.Exit(1)
 	}
 
+	kubeClient, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		setupLog.Error(err, "unable to create kubernets client")
+		os.Exit(1)
+	}
+
+	ctx := context.Background()
+	client, err := crclient.New(mgr.GetConfig(), crclient.Options{Scheme: scheme})
+	if err != nil {
+		setupLog.Error(err, "unable to create client")
+		os.Exit(1)
+	}
+
+	ok, err := utils.IsOpenShift(ctx, client)
+	if err != nil {
+		setupLog.Error(err, "unable to determine if cluster is OpenShift")
+		os.Exit(1)
+	}
+
+	if ok {
+		if err = (&controllers.ClusterVersionReconciler{
+			Client:     mgr.GetClient(),
+			Scheme:     mgr.GetScheme(),
+			KubeClient: kubeClient,
+		}).SetupWithManager(mgr); err != nil {
+			setupLog.Error(err, "unable to create controller", "controller", "ClusterVersionReconciler")
+			os.Exit(1)
+		}
+	} else {
+		err = templates.InitializeSidecars(kubeClient)
+		if err != nil {
+			setupLog.Error(err, "unable to initialize sidecars")
+			os.Exit(1)
+		}
+
+		err = csi.EnsureCSI(ctx, client, setupLog)
+		if err != nil {
+			setupLog.Error(err, "unable to create/update CSI")
+			os.Exit(1)
+		}
+	}
```

Comment:
In the standard kubernetes version, there is no clusterversion CR. AFAIK the server version will get updated first and during the kubelet upgrades on the nodes, the container will get restarted. the operator restart in this case, will read the server version and takes care of picking the right driver version. I was not sure about the upgrade in the OCP version and if the logic remains the same we can remove the controller for handling the csi resources.

Reply:
>since we have a periodic job running for the heartbeat, that should be taking care of the majority of things that we have to try and reconcile between upgrades, right

The heartbeat makes 2 RPC calls , the heartbeat job on the client side The heartbeat job is only to refresh the monitors IP and it will not,

between upgrades we only need to update the csi objects. let's keep it for now if theory works we can take the ontroller out :D

Reply:
Sounds good to me!